docs for keyset generation on anytrust orbits (#799)
OffchainLabs/arbitrum-docs: arbitrum-docs/launch-orbit-chain/concepts/anytrust-orbit-chain-keyset-generation.md

---
title: 'Anytrust orbit chains keyset generation'
sidebar_label: 'Keyset Generation for Anytrust Orbits'
description: 'Keyset generation and setup for Anytrust Orbit chains'
author: Mehdi Salehi
sme: Mehdi Salehi
target_audience: 'Developers deploying and maintaining Anytrust Orbit chains.'
sidebar_position: 1
---
**Anytrust** chains, relies on an external Data Availability Committee (hereafter, "the Committee") to store data and provide it on demand, instead of using the parent chain as Data Availability (DA) layer. 

> The Committee comprises N members, among whom it is presumed by AnyTrust that a minimum of two maintain integrity. Consequently, in scenarios where N - 1 members of the Committee pledge to grant access to specific data, it is guaranteed that at least one member among those committing to this promise upholds honesty.

Reviewer: The minimum number of trusted committee members is configurable by the "assumed-honest" parameter in the keyset.

Each member of the Committee is provisioned with a unique pair of BLS public and private keys. It is incumbent upon every member to independently generate a new, secure, and randomly derived set of BLS keys for their exclusive use. To generate a set of BLS keys, you can find more info [here](https://docs.arbitrum.io/node-running/how-tos/running-a-daserver#generate-key) on our docs.

> The parent chain is required to be cognizant of the composition of the Committee, encompassing the identity of each member along with their respective public keys, to facilitate verification processes. A 'Keyset' delineates the public keys of all Committee members and stipulates the requisite number of signatures for the legitimacy of a Data Availability Certificate. This configuration of Keysets not only facilitates modifications in the Committee's composition but also empowers members to update their individual key pairs as necessary. For more info, please read <a data-quicklook-from='inside-anytrust'>Inside Anytrust</a> page of our docs.

Reviewer: (style comment, feel free to ignore)
"cognizant", "encompassing", "facilitate", "delineate", "empower" 
I'd make the the language here simpler and more direct/specific.

<PublicPreviewBannerPartial />

In the following section, we will provide a detailed guide on the generation of a Keyset corresponding to your individual set of keys, as well as instructions for its subsequent configuration within the chain.

### Batch Poster Configuration

> AnyTrust uses a committee of Data Availability Servers to provide access to batched transaction data. The Nitro Batch Poster needs to be configured with, for each committee member, the URL, BLS public key of that committee member, and a single bit bitmask to uniquely identify that committee member, and a parameter called `assumed-honest` that sets how many committee members need to respond that they successfully stored the data. 

Reviewer: assumed-honest is H in the formula below, so it's true that it (indirectly) sets the number of committee members that need to respond, but it might be misinterpreted here that it directly sets it.

K = (N + 1) - H

K: Number of committee members that must sign for Store to be considered successful 
N: Total number of committee members
H: Minimum number of honest members

---
title: 'Anytrust orbit chains keyset generation'
sidebar_label: 'Keyset Generation for Anytrust Orbits'
description: 'Keyset generation and setup for Anytrust Orbit chains'
author: Mehdi Salehi
sme: Mehdi Salehi
target_audience: 'Developers deploying and maintaining Anytrust Orbit chains.'
sidebar_position: 1
---

> **Anytrust** chains, relies on an external Data Availability Committee (hereafter, "the Committee") to store data and provide it on demand, instead of using the parent chain as Data Availability (DA) layer. 

Reviewer 2: ```suggestion
**AnyTrust** chains rely on an external Data Availability Committee (DAC) to store data and provide it on demand, instead of using the parent chain as Data Availability (DA) layer. 
```

We use DAC elsewhere [[1]](https://docs.arbitrum.foundation/dao-glossary#data-availability-committee-dac); edit for consistency. This would also let us use quicklooks to expose term definitions (pulled from [Glossary CMS](https://www.notion.so/arbitrum/Glossary-CMS-0718756f84b44dcfa8907a72854550cf)) on hover.

> The Committee comprises N members, among whom it is presumed by AnyTrust that a minimum of H members maintain integrity. H is the minimum number of trusted committee members on Anytrust chains, and it is configurable by the "assumed-honest" parameter in the keyset, and by chain owner(s). Consequently, in scenarios where `K = (N + 1) - H` members of the Committee pledge to grant access to specific data, and they must sign and attest they have the data for store to be considered successful.

Reviewer 2: ```suggestion
The DAC has _N_ members; the AnyTrust protocol assumes that a minimum of _H_ DAC members maintain integrity. _H_ is the minimum number of trusted committee members on AnyTrust chains, configurable by the chain's owner via the `assumed-honest` parameter in the keyset. In scenarios where `K = (N + 1) - H` members of the DAC pledge to grant access to specific data, they must sign and attest they have the data for store to be considered successful.
```

Does this edit feel like an improvement?

> Consequently, in scenarios where `K = (N + 1) - H` members of the Committee pledge to grant access to specific data, and they must sign and attest they have the data for store to be considered successful.

I don't understand what we're trying to communicate here. Would you mind rearticulating / clarifying?

> Each Committee member gets their own set of BLS public and private keys. It's important for every member to create their own new and secure BLS keys. They should do this on their own and make sure these keys are random and only for their use. If you need help generating BLS keys, check out our guide [here](https://docs.arbitrum.io/node-running/how-tos/running-a-daserver#generate-key).

Reviewer 2: ```suggestion
Each DAC member gets their own set of BLS public and private keys. It's important for every member to create their own new and secure BLS keys. They should do this on their own and make sure these keys are random and only for their use. If you need help generating BLS keys, check out our guide [here](https://docs.arbitrum.io/node-running/how-tos/running-a-daserver#generate-key).
```

> The main blockchain (parent chain) needs to know who is in the Committee. It needs to have the names of all the members and their public keys. This is important for checking and verifying information. A 'Keyset' is a list of all the public keys of the Committee members. It also shows how many signatures are needed to approve a Data Availability Certificate. This Keyset setup lets us change who is in the Committee and lets members change their keys if needed. For more details, please read the <a data-quicklook-from='inside-anytrust'>Inside Anytrust</a> page in our documentation.

Reviewer 2: ```suggestion
The main blockchain (parent chain) needs to know the names and public keys of all DAC members in order to validate the integrity of data being batched and posted. A 'keyset' is a list of all DAC members' public keys. It also shows how many signatures are needed to approve a Data Availability Certificate. This design lets the chain owner modify the DAC's membership over time, and it lets DAC members change their keys if needed. See <a data-quicklook-from='inside-anytrust'>Inside Anytrust</a> for more information.
```

<PublicPreviewBannerPartial />

> In the following section, we will provide a detailed guide on the generation of a Keyset corresponding to your individual set of keys, as well as instructions for its subsequent configuration within the chain.

Reviewer 2: @GreatSoshiant how do you feel about refactoring the procedural instructions out of this document, and into a dedicated how-to document?

Our [established conventions](https://www.notion.so/arbitrum/Pattern-guide-Docs-dafaf8d1ce34433d88c1ef3b4b99c19c?pvs=4#2309f7f5db284126929065ac3c56fef8) distinguish between procedural and explanatory documents to make it easier for readers to find what they're looking for.

Not a blocker, but important if we want to maintain conventions (we can always address via another PR if this is time-sensitive).

If you opt to refactor now, Orbit docs auto-generates the sidebar from the directory structure, so creating a new how-to document in the `how-tos` directory should be all that's needed. We can also link to the how-to from the concept, and vice-versa, to improve discoverability of both.

Author: Can we stick to this type of doc for now @symbolpunk ? We can move to the how-to doc later if you agree.
